Input File Updates

gwenchee opened the pull request:
In this PR, I added 2 new input parameters to ROLLO's input file and updated one input parameter. 

Added parameters 
`weight`: list. This parameter was added into the algorithm section of the input file, and indicates the weight for each objective as a list. 
`order`: int. This parameter was added into the evaluators section of the input file, and indicates the order of software evaluations. It is indexed by 0 

Updated parameters 
`keep_files`: string. This parameter used to be a True, False Boolean. I updated it to be a string input. Accepted strings are `none`, `only_final`, `all`.

pep8speaks commented:
Hello @gwenchee! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





* In the file [`rollo/constraints.py`](https://github.com/arfc/rollo/blob/89f0be16a760df976b823eb64d02ab2e2d458400/rollo/constraints.py):

> [Line 124:80](https://github.com/arfc/rollo/blob/89f0be16a760df976b823eb64d02ab2e2d458400/rollo/constraints.py#L124): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (122 > 79 characters)

yardasol commented on rollo/evaluation.py:
```diff
@@ -125,12 +131,21 @@ def eval_function(ind):
 
         return eval_function
 
+    def solver_order(self, input_evaluators):
```
This needs a docstring, or at the minimum a comment explaining what it's purpose as a helper function is.

gwenchee replied:
Oh yes, i totally forgot to add the docstirng

yardasol commented on rollo/evaluation.py:
```diff
@@ -63,7 +63,12 @@ def add_evaluator(self, solver_name, input_script, output_script):
+            control_dict,
+            output_dict,
+            input_evaluators,
+            gens):
```
This parameter isn't listed in the docstring for this function.

gwenchee replied:
I also forgot to do this. Added it now.

yardasol reviewed (commented):
Hi @gwenchee, thanks for your contribution. These changes are quite clean! I'm having a bit of trouble understanding them though. Could you elaborate on the big picture purpose/motivation?

Also, I have a few comments:
1. I like that you're using jsonschema to validate your input files. However, I think defining the schema inside the python file takes up quite a bit of space and can distract from the focus on the algorithm of the functions. Consider moving the schema definitions to their own `.json` files and adding descriptions for each object/parameter as part of the schema.

2. Do you have unit tests for testing that the new input parameters are successfully validated?

- [x] Read the PR description
- [x] Read through all the changes, considering the following questions.
  - [x] Is there anything you can praise about this PR? Start with
        praise.
  - [x] Are variable names brief but descriptive?
  - [x] Are new/changed functions no longer than a paragraph?
  - [ ] Do all function parameters have default values where appropriate?
  - [x] Is the code clear and clean? (see Robert C. Martin's
        [Clean Code](https://i-share-uiu.primo.exlibrisgroup.com/discovery/fulldisplay?docid=alma99944155312205899&context=L&vid=01CARLI_UIU:CARLI_UIU&tab=LibraryCatalog&lang=en))
  - [x] Is there enough documentation?
  - [x] Does the programming style meet the requirements of the
        repository ([PEP8](https://www.python.org/dev/peps/pep-0008/) for python,
        [google C++ style guide](https://google.github.io/styleguide/cppguide.html), etc.)
  - [ ] If a new feature has been added, or a bug fixed, has a test been
        added to confirm good behavior?
  - [ ] Does the test actually test the new/changed functionality?
  - [ ] Does the test successfully test edge cases?
  - [ ] Does the test successfully test corner cases?
  - [ ] If the repository has continuous integration: does the PR pass
        the tests?
  - [ ] If there is no continuous integration, check out the branch
        locally, build, and run the tests.
  - [ ] Do the tests pass on your machine?
  - [x] Is the PR free of random cruft (built files, `.swp` files,
        etc.)?
  - [x] Do all files in the PR belong in the repository?
  - [x] Make a review, leaving kind comments and suggesting changes
        where needed (to resolve the above).
  - [x] Has the author resolved all of the comments and suggestions
        in your review?
- [ ] When you approve of the PR, merge and close it.
- [ ] Does this PR close an issue? If so, be sure to descriptively
      close this issue when the PR is merged
- [x] Thank the author for their contribution.

gwenchee commented:
Thanks for the review @yardasol. I addressed your comments by updating the docstrings. 

> Hi @gwenchee, thanks for your contribution. These changes are quite clean! I'm having a bit of trouble understanding them though. Could you elaborate on the big picture purpose/motivation?

In the big picture, the `weights` parameter allows the user to indicate the weighting of each objective which is interesting for multi-objective simulations. The `order` parameter is used to ensure that the evaluators are run in the correct order, and also as a sanity check for the user to feel more in control of the order of evaluations.  `keep_files` parameter used to be a boolean, but I added the `only_final` option for users that lack file space, but still want the option to see their evaluators files for the final generation. 

> Also, I have a few comments:
> 
> 1. I like that you're using jsonschema to validate your input files. However, I think defining the schema inside the python file takes up quite a bit of space and can distract from the focus on the algorithm of the functions. Consider moving the schema definitions to their own `.json` files and adding descriptions for each object/parameter as part of the schema.

I disagree with this. The schema validation happens in `input_validation.py` which is it's own contained file, not distracting from `algorithm.py` and `evaluation.py` where the actual logic takes place. 

> 2. Do you have unit tests for testing that the new input parameters are successfully validated?
I don't I added an issue for this. #29

gwenchee commented:
@yardasol bump

yardasol reviewed (approved):
Thanks for addressing my comments. I'm going to approve.

I want to clarify a bit my comment about the JSON Schema: when I said that having the schema defined in `input_validation.py` distracts from the focus of the algorithm, I was specifically referring to the functions in `input_validation.py`, since you seem to have a lot of infrastructure set up in that particular file. I was not referencing any other files in the repo. I do agree with your point that since everything in that file is related to input validation, it's not a huge issue to define the JSON schema in there.

Another thought I just had is that moving the JSON schema for the main ROLLO input file to its own `.json` file may make the input file structure a bit more discoverable than having it defined in `input_validation.py`. I think discoverability is important since input files are a user-facing component.

gwenchee commented:
@yardasol Yeah i totally see where you're coming from about the input_validation. I'll keep that in mind
